File: 11 Container With Most Water/C/solution.c
int min(int a, int b);
int max(int a, int b);

int maxArea(int* height, int heightSize) {
    int left = 0;
    int right = heightSize - 1;
    int maxWater = 0;
    while (left < right) {
        int water = min(height[left], height[right]) * (right - left);
        maxWater = max(maxWater, water);
        if (height[left] < height[right]) {
            left++;
        }
        else {
            right--;
        }
    }
    return maxWater;
}

int min(int a, int b) {
    return a <= b ? a : b;
}

int max(int a, int b) {
    return a >= b ? a : b;
}

File: 11 Container With Most Water/Python/Solution.py
from typing import List, Self


class Solution:
    def maxArea(self: Self, height: List[int]) -> int:
        left: int = 0
        right: int = len(height) - 1
        maxWater = 0
        while left < right:
            water: int = min(height[left], height[right]) * (right - left)
            maxWater = max(maxWater, water)
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        return maxWater